src/LifecycleWatcher.ts

```diff
@@ -14,18 +14,49 @@
  * limitations under the License.
  */
 
-const {helper, assert} = require('./helper');
-const {Events} = require('./Events');
-const {TimeoutError} = require('./Errors');
-
-class LifecycleWatcher {
-  /**
-   * @param {!Puppeteer.FrameManager} frameManager
-   * @param {!Puppeteer.Frame} frame
-   * @param {string|!Array<string>} waitUntil
-   * @param {number} timeout
-   */
-  constructor(frameManager, frame, waitUntil, timeout) {
+import {helper, assert, PuppeteerEventListener} from './helper';
+import {Events} from './Events';
+import {TimeoutError} from './Errors';
+
+type PuppeteerLifeCycleEvent = 'load' | 'domcontentloaded' | 'networkidle0' | 'networkidle2';
+type ProtocolLifeCycleEvent = 'load' | 'DOMContentLoaded' | 'networkIdle' | 'networkAlmostIdle';
+
+const puppeteerToProtocolLifecycle = new Map<PuppeteerLifeCycleEvent, ProtocolLifeCycleEvent>([
+  ['load', 'load'],
+  ['domcontentloaded', 'DOMContentLoaded'],
+  ['networkidle0', 'networkIdle'],
+  ['networkidle2', 'networkAlmostIdle'],
+]);
+
+
+export class LifecycleWatcher {
+  _expectedLifecycle: ProtocolLifeCycleEvent[];
+  _frameManager: Puppeteer.FrameManager;
+  _frame: Puppeteer.Frame;
+  _timeout: number;
+  _navigationRequest?: Puppeteer.Request;
+  _eventListeners: PuppeteerEventListener[];
+  _initialLoaderId: string;
+
+
+  _sameDocumentNavigationPromise: Promise<Error | null>;
+  _sameDocumentNavigationCompleteCallback: (x?: Error) => void;
+
+  _lifecyclePromise: Promise<void>;
+  _lifecycleCallback: () => void;
+
+  _newDocumentNavigationPromise: Promise<Error | null>;
+  _newDocumentNavigationCompleteCallback: (x?: Error) => void;
+
+  _terminationPromise: Promise<Error | null>;
+  _terminationCallback: (x?: Error) => void;
+
+  _timeoutPromise: Promise<TimeoutError | null>;
+
+  _maximumTimer?: NodeJS.Timeout;
+  _hasSameDocumentNavigation?: boolean;
```

Comment:
I don't like how many of these there are but decided to do the minimal amount of work to migrate; we can refactor and tidy once in TS land. Also keeps the PRs less risky and easier to review.